### PR TITLE
Allow tainting function arguments

### PR DESCRIPTION
**Pull request by JarLob**
This pull request allows tainting method arguments itself that is a common case in MVC pattern, like:
```cs
public class MyController : Controller
{
    public string DoSomething(string input) // << input is tainted argument
    {
        new SqlCommand(input);
        return null;
    }
}
```

### CODE REVIEW

**Comment by dotpaul** [comment]
Thank you again, @JarLob ! Could you target the 2.9.x branch also? For the time being, we're still releasing 2.9.x packages with security rule fixes and additions.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataSymbolMapExtensions.cs [hunk 1]
```diff
@@ -130,6 +130,24 @@ public static bool IsSourceProperty(this TaintedDataSymbolMap<SourceInfo> source
...
+        /// <param name="sourceSymbolMap"></param>
+        /// <param name="parameterSymbol"></param>
+        /// <returns></returns>
+        public static bool IsSourceParameter(this TaintedDataSymbolMap<SourceInfo> sourceSymbolMap, IParameterSymbol parameterSymbol)
+        {
+            ISymbol containingSymbol = parameterSymbol.ContainingSymbol;
+            foreach (SourceInfo sourceInfo in sourceSymbolMap.GetInfosForType(containingSymbol.ContainingType))
+            {
+                if (sourceInfo.TaintedArguments.Any(match => match(parameterSymbol)))
+                    return true;
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataSymbolMapExtensions.cs:145]:
```suggestion
                {
                    return true;
                }
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs [hunk 2]
```diff
@@ -19,6 +21,21 @@ static WebInputSources()
...
             var sourceInfosBuilder = PooledHashSet<SourceInfo>.GetInstance();
 
+            sourceInfosBuilder.AddSourceInfo(
+                WellKnownTypeNames.MicrosoftAspNetCoreMvcControllerBase,
+                 new ParameterMatcher[]{
+                    (parameter) => {
+                        ISymbol containingSymbol = parameter.ContainingSymbol;
+
+                        if (containingSymbol.DeclaredAccessibility != Accessibility.Public)
+                            return false;
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:31]:
```suggestion
                        {
                            return false;
                        }
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs [hunk 3]
```diff
@@ -19,6 +21,21 @@ static WebInputSources()
         {
             var sourceInfosBuilder = PooledHashSet<SourceInfo>.GetInstance();
 
+            sourceInfosBuilder.AddSourceInfo(
+                WellKnownTypeNames.MicrosoftAspNetCoreMvcControllerBase,
+                 new ParameterMatcher[]{
+                    (parameter) => {
+                        ISymbol containingSymbol = parameter.ContainingSymbol;
+
+                        if (containingSymbol.DeclaredAccessibility != Accessibility.Public)
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:30]:
```suggestion
                        if (containingSymbol.DeclaredAccessibility != Accessibility.Public
                            || containingSymbol.IsConstructor())
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs [hunk 4]
```diff
@@ -19,6 +21,21 @@ static WebInputSources()
...
+                            return false;
+
+                        if (containingSymbol.IsConstructor())
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:33]: (Then this `if` could be removed)

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs [hunk 5]
```diff
@@ -19,6 +21,21 @@ static WebInputSources()
         {
             var sourceInfosBuilder = PooledHashSet<SourceInfo>.GetInstance();
 
+            sourceInfosBuilder.AddSourceInfo(
+                WellKnownTypeNames.MicrosoftAspNetCoreMvcControllerBase,
+                 new ParameterMatcher[]{
+                    (parameter) => {
+                        ISymbol containingSymbol = parameter.ContainingSymbol;
+
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:29]: Definitely should take into account `[NonAction]` attributes, and possibly `[NonController]` attributes. We can't have the `INamedTypeSymbol`s at this point, so may have to do something like look at the `MetadataName` of `AttributeData.AttributeClass`, and walk up their `ContainingNamespace`s checking their `MetadataName`s.

https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/actions?view=aspnetcore-3.1 suggests that controllers don't necessarily have to derive from `ControllerBase`, but that's probably the most common case.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs (hunk 3 above)
Comment (dotpaul) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:30]: Also, please add unit tests for these cases.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs (hunk 4 above)
Comment (JarLob) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:33]: Do you mean `return !containingSymbol.IsConstructor();`?

Comment (JarLob) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:33]: Ah, I see
```cs
if (containingSymbol.DeclaredAccessibility != Accessibility.Public
                            || containingSymbol.IsConstructor())
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs (hunk 5 above)
Comment (JarLob) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs:45]: Love the document, is there something like this for the old asp.net mvc?

**Comment by JarLob** [comment]
> ASP.NET (non-Core) MVC controllers would be neat too (but don't feel obligated to add 'em).

Every time I need to add a new reference in unit tests it is a pain. I'll you guys implement it ;)

**Comment by JarLob** [comment]
> Thank you again, @JarLob ! Could you target the 2.9.x branch also? For the time being, we're still releasing 2.9.x packages with security rule fixes and additions.

As the last time? Close this PR open a new one? Or commit in both branches?

**Comment by JarLob** [comment]
Since you are here. Is there a way to define a `return` from such method as a `sink`? Like when it returns a non html encoded string.

**Comment by dotpaul** [comment]
> As the last time? Close this PR open a new one? Or commit in both branches?

Close this PR, open a new one. Or rebase and force push to your fork, and retarget this PR. Definitely only submit a PR to the 2.9.x branch.

> Every time I need to add a new reference in unit tests it is a pain. I'll let you guys implement it ;)

Fair enough. 😄 

> Since you are here. Is there a way to define a `return` from such method as a `sink`? Like when it returns a non html encoded string.

Can you provide an example?

**Comment by JarLob** [comment]
> Can you provide an example?

```cs
public class MyController : Controller
{
    public string DoSomething(string input) 
    {           
        return input; // << input is tainted, treat return as sink
    }
}
```

**Comment by dotpaul** [comment]
Closing in favor of #3966. Continuing discussion of return values as sinks in #3990.

**Pull request closed without merging**